grub-core/loader/multiboot_elfxx.c

```diff
@@ -101,30 +111,99 @@ CONCAT(grub_multiboot_load_elf, XX) (mbi_load_data_t *mld)
     {
       load_size = highest_load - mld->link_base_addr;
 
-      grub_dprintf ("multiboot_loader", "align=0x%lx, preference=0x%x, "
+      grub_dprintf ("multiboot_loader", "link_base_addr=0x%08x, align=0x%lx, preference=0x%x, "
 		    "load_size=0x%x, avoid_efi_boot_services=%d\n",
-		    (long) mld->align, mld->preference, load_size,
+		    mld->link_base_addr, (long) mld->align, mld->preference, load_size,
 		    mld->avoid_efi_boot_services);
 
-      if (load_size > mld->max_addr || mld->min_addr > mld->max_addr - load_size)
+      if (grub_slaunch_platform_type () == SLP_INTEL_TXT)
+	{
+#ifndef GRUB_USE_MULTIBOOT2
+	  return grub_error (GRUB_ERR_BAD_OS, "Only multiboot2 supported for slaunch");
+#else
+	  /*
+	   * We allocate the the binary together with page tables and MBI
+	   * to make one contiguous block for MLE. We have to align up to
+	   * PMR (2MB).
+	   */
+	  total_size = ALIGN_UP(load_size, 8); // MULTIBOOT_TAG_ALIGN
+	  total_size += GRUB_MULTIBOOT (get_mbi_size)();
+	  total_size = ALIGN_UP(total_size, GRUB_TXT_PMR_ALIGN);
+
+	  slparams->mle_size = total_size;
```

Comment:
We probably shouldn't measure MBI together with code.
```suggestion
	  slparams->mle_size = load_size;
```

Reply:
Then how can we trust what was passed here? I.e. the kernel and initrd and other params? IIRC the Xen cmdline is passed via MBI2 too, so it can't be ignored.

Reply:
Xen should measure it to different PCR (to isolate code from configuration) before use, just as Linux does (or should do, it lacks "before use" part) with zero page. There is another task created to develop this.

I am aware that this will be different from approach taken on AMD platforms, but since we can't change how SINIT ACM works, our options are limited.

Reply:
Okay so I will need to rewrite a bit of it... Anyways I need to keep a record of where MBI is...

Reply:
So are you going to measure MBI in the assembly in Xen?

Reply:
If you did review https://github.com/3mdeb/xen/pull/11 you would know that it can be done in C :slightly_smiling_face:

Reply:
Linux implementation has the mle_size aligned to PMR align (2MB), so I will simply leave load_size aligned up to 2MBs here without MBI.

Reply:
Now MBI is not part of MLE

Reply:
By having MLE size aligned to 2MB we're wasting time measuring empty space. In fact, I don't see why this have to be aligned since PMR covers whole memory and I haven't found anything in TXT docs that would require such alignment.

Reply:
This is what Linux implementation does, I have simply replicated same approach.

Reply:
OK, until we get some confirmation as to why it is done this way, we can leave it as it is.

I still can't get expected MLE measurements, when compared with (uncompressed) file contents, no matter which alignment is used. Padding should be cleared to 0, but can't tell for sure. I don't feel like dumping 4MB of memory through serial to compare it, but I may have to... Another option is that there is something wrong with page tables and something else is measured, but I doubt it, after all it jumps into proper place in Xen later, which is also based on the same page tables.

Reply:
Just uncompressing isn't enough. Do you strip the ELF header? Do you calculate measurement based on the loadable section only?

Reply:
Obviously. I use something like this:

```
$ objdump xen -p

xen:     file format elf32-i386

Program Header:
    LOAD off    0x00000080 vaddr 0x00200000 paddr 0x00200000 align 2**6
         filesz 0x002c2900 memsz 0x003f8000 flags rwx
    NOTE off    0x00206858 vaddr 0x002067d8 paddr 0x002067d8 align 2**2
         filesz 0x00000024 memsz 0x00000024 flags r--

$ dd if=xen ibs=1 skip=$((0x80)) count=$((0x2c2900)) obs=$((0x3f8000)) conv=sync,notrunc | dd ibs=$((0x3f8000)) conv=sync,notrunc | sha1sum
```

This is Xen with additional debug lines added so your sizes will probably be different.

Reply:
Anyway no idea what is going on.

Reply:
OK, I'm marking it as resolved, since I don't see anything wrong with the way it is configured. Aligning everything to PMR may be an overkill, but it works. We may get back to it if/when something breaks for newer platforms, e.g. Daniel mentioned that page tables have to live directly below MLE, with no padding, which isn't true for us at the moment.